### PR TITLE
[ROCM] MIOpen contrib for convolution kernels

### PR DESCRIPTION
**Pull request by masahi**
There are several differences with cudnn contrib:
* MIOpen compiles and caches kernel codes during runtime
* Compilation happens during mandatory miopenFindConvolutionForwardAlgorithm(...)
* miopenFindConvolutionForwardAlgorithm(...) runs a benchmark to find the best algo for the provided input. 
* To avoid running miopenFindConvolutionForwardAlgorithm on every extern call, I run it once during build time. Workspace is also setup during build time. 

I attached a test case which compares miopen output with TVM AMDGPU backend.
```
$ python test_miopen.py 
[22:50:44] src/contrib/miopen/conv_forward.cc:136: 	MIOpen Found 1 fwd algorithms, choosing miopenConvolutionFwdAlgoWinograd
[22:50:44] src/contrib/miopen/conv_forward.cc:139: 		0) miopenConvolutionFwdAlgoWinograd - time: 0.110878 ms, Memory: 0
/home/masa/project/dev/tvm/python/tvm/build_module.py:340: UserWarning: Specified target rocm, but cannot find device code, did you do bind?
  "Specified target %s, but cannot find device code, did you do bind?" % target)
Max abs diff: 4.19617e-05
```
@adityaatluri

### CODE REVIEW

File: src/contrib/miopen/miopen_utils.h [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+std::string miopenGetErrorString(int error_code);
+
+#define MIOPEN_CALL(func)                                                      \
+  {                                                                            \
+    miopenStatus_t e = (func);                                                 \
+    CHECK_EQ(e, miopenStatusSuccess)                                           \
+        << "miopen error: " << miopenGetErrorString(e);                        \
+  }
+
+struct ConvEntry {
```
Comment (tqchen) [inline src/contrib/miopen/miopen_utils.h]: There might be namespace issue when compile with other contrib with same conventry name, consider add a namespace miopen here

Comment (masahi) [reply src/contrib/miopen/miopen_utils.h]: done

**Comment by masahi** [comment]
@tqchen Don't merge this yet. I think I found a bug.

**Comment by masahi** [comment]
@tqchen I fixed my bug. Now MIOpen works with NNVM.
 I can run resnet and vgg benchmark with MIOpen kernels.

**Comment by masahi** [comment]
@tqchen I added "use_miopen" to tvm.target.Target. See my last commit.

**Comment by masahi** [comment]
@tqchen can you check my change is legit?

**Comment by tqchen** [comment]
@masahi we can update LLVM's option detector to simply skip this option, so it won;t result in a fault

**Comment by masahi** [comment]
@tqchen Sorry, I found a typo

**Comment by masahi** [comment]
fixed

**Comment by tqchen** [comment]
try to run test locally, they might have things to do with your change in target

**Comment by masahi** [comment]
ok, I can reproduce errors locally. 
will fix

**Comment by masahi** [comment]
fixed

**Comment by masahi** [comment]
@tqchen jenkins is failing on one of the topi unittests.
I can pass test_topi_reduce.py locally.

**Comment by tqchen** [comment]
OK, that one is just fine, sometimes reduction tests fail possibly due to a tie. nd won't do so after retrigger

**Comment by masahi** [comment]
it's failing again.

**Comment by masahi** [comment]
Test passed. Ready to merge.

**Pull request merged**